src/agents/experimental/streamlit/agent.ts

```diff
@@ -0,0 +1,182 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { AgentError, BaseAgent, BaseAgentRunOptions } from "@/agents/base.js";
+import { AgentMeta } from "@/agents/types.js";
+import { GetRunContext } from "@/context.js";
+import { Callback, Emitter } from "@/emitter/emitter.js";
+import { BaseMemory } from "@/memory/base.js";
+import { ChatLLM, ChatLLMOutput } from "@/llms/chat.js";
+import { isTruthy } from "remeda";
+import { BaseMessage, Role } from "@/llms/primitives/message.js";
+import {
+  StreamlitAgentSystemPrompt,
+  StreamlitAgentTemplates,
+} from "@/agents/experimental/streamlit/prompts.js";
+import { BaseLLMOutput } from "@/llms/base.js";
+import { TokenMemory } from "@/memory/tokenMemory.js";
+import { findFirstPair } from "@/internals/helpers/string.js";
+
+export interface StreamlitAgentInput {
+  llm: ChatLLM<ChatLLMOutput>;
+  memory: BaseMemory;
+  templates?: Partial<StreamlitAgentTemplates>;
+}
+
+export interface StreamlitRunInput {
+  prompt: string | null;
+}
+
+interface Options extends BaseAgentRunOptions {}
+
+interface Result {
+  raw: string;
+
+  app: string | null;
+  text: string;
+}
+
+export interface StreamlitRunOutput {
+  result: Result;
+  message: BaseMessage;
+  memory: BaseMemory;
+}
+
+export interface StreamlitEvents {
+  newToken: Callback<{
+    delta: string;
+    state: Readonly<{
+      content: string;
+    }>;
+    chunk: BaseLLMOutput;
+  }>;
+}
+
+export class StreamlitAgent extends BaseAgent<StreamlitRunInput, StreamlitRunOutput, Options> {
+  public emitter = new Emitter<StreamlitEvents>({
+    namespace: ["agent", "experimental", "streamlit"],
+    creator: this,
+  });
+
+  constructor(protected readonly input: StreamlitAgentInput) {
+    super();
+  }
+
+  public get meta(): AgentMeta {
+    return {
+      name: `Streamlit`,
+      tools: [],
+      description: `StreamlitAgent is an experimental meta-app agent that uses \`Meta LLaMa 3.1 70B\` to build \`IBM Granite 3 8B\`-powered apps using Streamlit -- a simple UI framework for Python.`,
+    };
+  }
+
+  public get memory() {
+    return this.input.memory;
+  }
+
+  protected async _run(
+    input: StreamlitRunInput,
+    _options: Options | undefined,
+    run: GetRunContext<typeof this>,
+  ): Promise<StreamlitRunOutput> {
+    const { userMessage, runMemory } = await this.prepare(input);
+
+    let content = "";
+    for await (const chunk of this.input.llm.stream(runMemory.messages, {
+      signal: run.signal,
+    })) {
+      const delta = chunk.getTextContent();
+      content += delta;
+      await run.emitter.emit("newToken", { delta, state: { content }, chunk });
+    }
+    const result = this.parse(content);
```

Comment:
This is quite unwieldy when the consumer structures the code mostly around emitter (like bee-api does). Would it be possible to have an event emitted or at least make the parse method public?

Reply:
There is no reason to make `parse` public. You would not get any beneficial data during streaming, and once executions end, you will get the parsed (and raw) result.